common/speedrunning.py

```diff
@@ -0,0 +1,259 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import datetime
+from datetime import timedelta
+from enum import StrEnum
+from uuid import UUID
+
+from objects import glob
+
+
+class ScoreType(StrEnum):
+    """The type of score that the user is speedrunning."""
+
+    # TODO: implicitly only loved/ranked/approved maps
+    WEIGHTED_PP = "weighted_pp"
+    WEIGHTED_SCORE = "weighted_score"
+
+
+class SpeedrunTimeframe(StrEnum):
+    """The timeframe that the user is speedrunning."""
+
+    TEN_MINUTES = "10m"
+    ONE_HOUR = "1h"
+    ONE_DAY = "1d"
+    ONE_WEEK = "1w"
+
+
+@dataclass
+class UserSpeedrun:
+    id: UUID
+    user_id: int
+    game_mode: int
+    timeframe: SpeedrunTimeframe
+    score_type: ScoreType
+    score_value: int
+    started_at: datetime
+    ended_at: datetime | None
+    cancelled_at: datetime | None
```

Comment:
todo rm this, can virtually figure this out `if ended_at and (ended_at - started_at) < timeframe`